### PR TITLE
Fixes #2358 - BREAKING CHANGE: `Pos.Combine` is incorrect for scenarios involving `PosAbsolute`

### LINKED ISSUE #2358: `Pos.Combine` is incorrect for scenarios involving `PosAbsolute`
Change the `ComputedScenario` to add an additional button like this:

```cs
                       Win.Add (rightButton);
....
			centerButton = new Button ("0 + Center") {
				X = 0 + Pos.Center (),
				Y = Pos.AnchorEnd (2)
			};
			Win.Add (centerButton);
```

Expected:

The new button is centered in the view.

Actual:

image

The bug is in `View.SetRelativeLayout`. It needs to check for `PosCombine` and iterate through `PosCombine.left` and `PosCombine.right` to adjust the relative location properly. 

I'm not actually sure it is possible to deal with all combinations. This may not be worth fixing.

Related: 

* What would you expect `Pos.Center() + Pos.Center()` to do?
*  What would you expect `Pos.Percent() + Pos.Center()` to do?
* How about `Pos.Percent(20) + Pos.Percent(20)`? (this one's obvious: just like `Pos.Percent(40)`).
etc...

### PR DESCRIPTION
**Pull request by tig**
* Fixes #2358 
* Prepares for v2 by making `SetRelativeLayout` less fragile
* In order to fix #2356 the following is part of this PR:
  * `SetRelativeLayout` and the related methods were refactored
  * Layout related unit tests were moved to `Core/LayoutTests.cs`
  * Adds unit tests to show #2358 
  * Fixes a bug in `Border` where it was not checking for negative row/col values

This is a breaking change (it changes the previous behavior of SetRelativeLayout) and breaks any code that was using `PosCombine` but not realizing the results were invalid. For example, all of the `Border` related Scenarios.

We should take this breaking change because it fixes a real problem in how the API is supposed to work.

## Pull Request checklist:

- [x] I've named my PR in the form of "Fixes #issue. Terse description."
- [x] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [x] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [x] I conducted basic QA to assure all features are working

### CODE REVIEW

**Comment by tig** [comment]
@BDisp - can you please do a CR of this?

**Comment by BDisp** [comment]
@tig I think you may merge the #2345 into this to leverage the frame border color not equal to the color scheme.

**Comment by BDisp** [comment]
@tig none of theses labels are redrawing on the `Borders Comparisons`, `Borders on FrameView`, `Borders on Toplevel` and `Borders on Window` scenarios.
```cs
var label = new Label ("I'm a Window") {
				X = Pos.Center (),
				Y = Pos.Center () - 3,
			};
var label2 = new Label ("I'm a Toplevel") {
				X = Pos.Center (),
				Y = Pos.Center () - 3,
			};
var label3 = new Label ("I'm a FrameView") {
				X = Pos.Center (),
				Y = Pos.Center () - 3,
			};
```

imagem

**Comment by BDisp** [comment]
On load theses to views are misaligned:
imagem

On the develop branch they are aligned:
imagem

I think this has to do with the `Pos.Center` and `Pos.Absolute` layout.

File: UnitTests/Core/BorderTests.cs [hunk 1]
```diff
@@ -1,19 +1,15 @@
...
-		readonly ITestOutputHelper output;
-
-		public BorderTests (ITestOutputHelper output)
-		{
-			this.output = output;
-		}
-
-		[Fact, AutoInitShutdown]
+		[Fact]
+		[AutoInitShutdown]
```
Comment (BDisp) [inline UnitTests/Core/BorderTests.cs:12]: Sorry but I really don't understand what is wrong here. Don't you like them in one line? `Ctrl+K, Ctrl+D` don't modify them, so you spend a lot to change them :-)

File: UnitTests/Core/BorderTests.cs [hunk 2]
```diff
@@ -557,57 +557,5 @@ public void BorderOnControlWithNoChildren ()
 
 			Assert.Null (Record.Exception (() => label.Redraw (label.Bounds)));
 		}
-
-		[Fact, AutoInitShutdown]
-		public void BorderStyle_And_DrawMarginFrame_Gets_Sets ()
```
Comment (BDisp) [inline UnitTests/Core/BorderTests.cs:562]: I hope you also maintain this unit test. I think you have some errors because of the `View.Clear` method.

**Review by BDisp — CHANGES_REQUESTED** [review]
I wrote some considerations about this. I hope you don´t mind.

**Comment by tig** [comment]
> On load theses to views are misaligned: imagem
> 
> On the develop branch they are aligned: imagem
> 
> I think this has to do with the `Pos.Center` and `Pos.Absolute` layout.

Yep, I struggled getting them to align using `Pos.Bottom(label)` but Panel's Frame is at the marginframe and label's is at the Margin so it wasn't working. I decided to leave it like this so that would be obvious (but I was also tired of fixing bugs in those Scenarios that were exposed by the bug fix).

File: UnitTests/Core/BorderTests.cs (hunk 1 above)
Comment (tig) [reply UnitTests/Core/BorderTests.cs:12]: Was a result of a merge, I think. I didn't do it on purpose.

File: UnitTests/Views/ScrollViewTests.cs [hunk 3]
```diff
@@ -276,227 +280,5 @@ public void ContentOffset_ContentSize_AutoHideScrollBars_ShowHorizontalScrollInd
 ◄░░░├─┤░► 
 ", output);
 		}
-
-		[Fact, AutoInitShutdown]
-		public void Frame_And_Labels_Does_Not_Overspill_ScrollView ()
```
[the remark this answers is not in the document]
Comment (tig) [reply UnitTests/Views/ScrollViewTests.cs:281]: They were moved to LayoutTest.cs (or at least I meant to move them... Or I deleted them by accident). I'll re-look.

**Comment by tig** [comment]
> @tig I think you may merge the #2345 into this to leverage the frame border color not equal to the color scheme.

Done. Thanks for noting that!

**Comment by tig** [comment]
> @tig none of theses labels are redrawing on the `Borders Comparisons`, `Borders on FrameView`, `Borders on Toplevel` and `Borders on Window` scenarios.
> 
> ```cs
> var label = new Label ("I'm a Window") {
> 				X = Pos.Center (),
> 				Y = Pos.Center () - 3,
> 			};
> var label2 = new Label ("I'm a Toplevel") {
> 				X = Pos.Center (),
> 				Y = Pos.Center () - 3,
> 			};
> var label3 = new Label ("I'm a FrameView") {
> 				X = Pos.Center (),
> 				Y = Pos.Center () - 3,
> 			};
> ```
> 
> imagem

This is caused by bugs in those scenarios that I didn't notice:

```cs
			var label = new Label ("I'm a Window") {
				X = Pos.Center (),
				Y = Pos.Center () - 3,
			};
```

`Y.PosCenter() - 3` for a 7 row tall view in a 5 row tall superview puts the label outside of the Bounds.

This works better:

```cs
			var label = new Label ("I'm a Window") {
				X = Pos.Center (),
				Y = Pos.Top (button) - 1
			};
``` 

image

This regex search finds all instances of potential bugs that do this `Pos.Center \(\) -`.

**Comment by tig** [comment]
I just pushed several fixes to this PR that fix the issues you noted. However, one fix that SHOULD have worked has caused another problem in the 3 "Borders on XXX" scenarios:

"Borders on Window"
image

"Borders on FrameView"
image

"Borders on Toplevel"
image

Since this doesn't happen in "Borders" I'm wondering if this is another bug in the Scenario. Debugging now...

The button isn't painting right.

**Comment by BDisp** [comment]
> Since this doesn't happen in "Borders Comparsion" I'm wondering if this is another bug in the Scenario. Debugging now...

I already noted that. Try to comment in the View on the GetMaxNeedDisplay method these two lines:
```diff
+			//rect.Width += Math.Max (oldFrame.X - newFrame.X, 0);
+			//rect.Height += Math.Max (oldFrame.Y - newFrame.Y, 0);

```

@tig I prefer to use the `Y = Pos.Center () - 2`, to really test that center is working properly. Change to -2, because you fixed some existent bug on the layout.

**Comment by tig** [comment]
> > Since this doesn't happen in "Borders Comparsion" I'm wondering if this is another bug in the Scenario. Debugging now...
> 
> I already noted that. Try to comment in the View on the GetMaxNeedDisplay method these two lines:
> 
> ```diff
> +			//rect.Width += Math.Max (oldFrame.X - newFrame.X, 0);
> +			//rect.Height += Math.Max (oldFrame.Y - newFrame.Y, 0);
> ```
> 
> @tig I prefer to use the `Y = Pos.Center () - 2`, to really test that center is working properly. Change to -2, because you fixed some existent bug on the layout.

The bug is here, I think:

image

`GetNeedDisplay` is `{{X=0,Y=0,Width=65,Height=17}}` for the label, which has `AutoSize` on and thus a `Height` of `. Thus this call to Clear is clearing more than it should:

image

I'm out of time for tonight. I you get a chance can you file a new issue for that?

**Comment by tig** [comment]
> @tig I prefer to use the Y = Pos.Center () - 2, to really test that center is working properly. Change to -2, because you fixed some existent bug on the layout.

Well, that didn't work, did it? Center WASN'T working right ;-).

I'm out of time tonight so I'm going to leave it as-is for now.

**Comment by tig** [comment]
Now that I think about it more: Why isn't `Clear` honoring `Driver.Clip`???

**Comment by BDisp** [comment]
> Now that I think about it more: Why isn't Clear honoring Driver.Clip???

If the Driver.Clip is equal to the current view bounds it's ok, otherwise is better maintain as is.

**Comment by tig** [comment]
I now no longer want to fix this in v1.

This bug has persisted for years and I'm the first to notice it. 

I am getting worn out trying to merge v1 changes into v2. I spent 3 hours yesterday  and ended up giving up because I had made such a mess.

**Comment by BDisp** [comment]
@tig I added the PR #2385 on the v2 that fixes the #2358. The behavior with the button redraw doesn't reproduce anymore. I hope you don't mind.

**Comment by tig** [comment]
Closing as a dupe of #2385

**Pull request closed without merging**